### DISCUSSION TITLE
typing for session factory pattern

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#11280] [Usage Questions] [answered]
I can't seem to figure out the right way to type the following code.

I get the following error:

> sqlalchemy_typing_test.py:19: error: "Session" not callable  [operator]

The callable pattern of `Session()` is used historically, and widely in the docs (see https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-frequently-asked-questions)

Am I doing something very wrong here?

```!python
from typing import Any
from typing import Callable
from typing import TYPE_CHECKING
from typing import Union

if TYPE_CHECKING:
    from pyramid.request import Request
    from sqlalchemy.orm import scoped_session
    from sqlalchemy.orm.session import Session
    import transaction
    _TYPES_SESSION = Union[Session, scoped_session[Any], Callable]


def get_tm_session(
    request: "Request",
    session_factory: "_TYPES_SESSION",
    transaction_manager: "transaction.manager",
):
    dbSession = session_factory(info={"request": request})
    return dbSession
```

**Comment by Alc-Alc** [comment, marked as answer]
I think you probably meant `_TYPES_SESSION = Union[type[Session], scoped_session[Any]]`.

Note the usage of `type` in `type[Session]`. IIUC you are passing in a session class not an instance of session itself.

As an aside, in Python < 3.9, you can't use type[Session], rather `typing.Type[Session]`

**Reply by jvanasco** [reply]
Thank you so much! That is the correction I needed!

**Reply by jvanasco** [reply]
It did need this slight change:

    _TYPES_SESSION = Union[Type[Session], scoped_session[Any], sessionmaker[Session]]]
